### PR TITLE
Statistic Update for RtpsRelay

### PR DESCRIPTION
**Pull request by jrw972**
* Adds a topic for Participants
* Adds a separate topic for Participant Statistics
* Removes "monitor mode"
* All statistics are written p.r.n.

### CODE REVIEW

**Comment by brattonj-oci** [comment]
OK, looks fine though it looks like you changed port numbers. Any specific reason?

**Review by brattonj-oci — APPROVED** [review]

File: tools/rtpsrelay/Config.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  Config()
+    : statistics_interval_(60) // 1 minute
+    , application_participant_guid_(OpenDDS::DCPS::GUID_UNKNOWN)
+    , lifespan_(60) // 1 minute
+    , application_domain_(1)
+    , publish_relay_statistics_(true)
+    , log_relay_statistics_(false)
+  {}
+
+  void statistics_interval(const OpenDDS::DCPS::TimeDuration& flag)
```
Comment (mitza-oci) [inline tools/rtpsrelay/Config.h:21]: Is there a name other than `flag` we could use here?

File: tools/rtpsrelay/DomainStatisticsReporter.h [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+#include "lib/RelayTypeSupportImpl.h"
+
+#include <iostream>
+
+namespace RtpsRelay {
+
+class DomainStatisticsReporter {
+public:
+  DomainStatisticsReporter(const Config& config,
+                           DomainStatisticsDataWriter_ptr writer)
```
Comment (mitza-oci) [inline tools/rtpsrelay/DomainStatisticsReporter.h:17]: Could this use `_var`?

File: tools/rtpsrelay/DomainStatisticsReporter.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    : config_(config)
+    , last_report_(OpenDDS::DCPS::MonotonicTimePoint::now())
+    , writer_(writer)
+  {
+    domain_statistics_.application_participant_guid(repoid_to_guid(config_.application_participant_guid()));
+    domain_statistics_.local_participants(0);
+    domain_statistics_.local_writers(0);
+    domain_statistics_.local_readers(0);
+    domain_statistics_.total_writers(0);
+    domain_statistics_.total_readers(0);
```
Comment (mitza-oci) [inline tools/rtpsrelay/DomainStatisticsReporter.h:27]: seems redundant

File: tools/rtpsrelay/ParticipantListener.h [hunk 4]
```diff
@@ -2,17 +2,26 @@
...
+#include "DomainStatisticsReporter.h"
+
+#include <dds/DCPS/DomainParticipantImpl.h>
 
 namespace RtpsRelay {
 
 class ParticipantListener : public ListenerBase {
 public:
-  explicit ParticipantListener(DomainStatisticsWriter& stats_writer);
+  explicit ParticipantListener(OpenDDS::DCPS::DomainParticipantImpl* participant,
```
Comment (mitza-oci) [inline tools/rtpsrelay/ParticipantListener.h:13]:
```suggestion
  ParticipantListener(OpenDDS::DCPS::DomainParticipantImpl* participant,
```

File: tools/rtpsrelay/ParticipantListener.h [hunk 5]
```diff
@@ -2,17 +2,26 @@
...
+                               DomainStatisticsReporter& stats_reporter,
+                               ParticipantEntryDataWriter_ptr participant_writer);
```
Comment (mitza-oci) [inline tools/rtpsrelay/ParticipantListener.h:15]: smart pointers preferred for both of these

File: tools/rtpsrelay/ParticipantStatisticsReporter.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  void output_message(size_t byte_count, const OpenDDS::DCPS::MonotonicTimePoint& now)
+  {
+    participant_statistics_._bytes_out += byte_count;
+    ++participant_statistics_._messages_out;
+    report(now);
+  }
+
+  void max_fan_out(size_t value, const OpenDDS::DCPS::MonotonicTimePoint& now)
+  {
+    participant_statistics_._max_fan_out = std::max(participant_statistics_._max_fan_out, static_cast<uint32_t>(value));
```
Comment (mitza-oci) [inline tools/rtpsrelay/ParticipantStatisticsReporter.h:45]:
```suggestion
    participant_statistics_.max_fan_out() = std::max(participant_statistics_.max_fan_out(), static_cast<uint32_t>(value));
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Comment by jrw972** [comment]
> OK, looks fine though it looks like you changed port numbers. Any specific reason?

I found out that the IANA has those ports set aside for RTPS traffic.  https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?skey=8&page=102

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**